### PR TITLE
Verify charmhub requests

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following introduces the validation process against the charmhub
API. We use two different validators to check if an upgrade series is
allowed via an application. Previously this was all done in state and now
we use the charmhub API to verify that for us.

A model can contain two different type of charms, a state charm (local
and charmstore charm) and a charmhub charm. To ensure that we don't
create a heterogeneous deployment we use a check to ensure that we call
the right one.

## QA steps

```sh
$ juju bootstrap lxd test
$ juju deploy ubuntu --series=bionic
$ juju upgrade-series 0 prepare focal
```

### CODE REVIEW

**Review by manadart — APPROVED** [review]
Looks good.

File: apiserver/facades/client/machinemanager/upgrade_series.go [hunk 1]
```diff
@@ -217,13 +227,61 @@ func (s upgradeSeriesState) ApplicationsFromMachine(machine Machine) ([]Applicat
...
 
-type upgradeSeriesValidator struct{}
+type upgradeSeriesValidator struct {
+	stateValidator   ApplicationValidator
+	requestValidator ApplicationValidator
+}
+
+func makeUpgradeSeriesValidator(client CharmhubClient) upgradeSeriesValidator {
+	return upgradeSeriesValidator{
+		stateValidator: stateSeriesValidator{},
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/facades/client/machinemanager/upgrade_series.go:237]: Ha, I had "legacy" as the original name!

File: apiserver/facades/client/machinemanager/upgrade_series.go [hunk 2]
```diff
@@ -249,3 +307,48 @@ func (s upgradeSeriesValidator) verifySupportedSeries(application Application, s
...
+	refreshResp, err := s.client.Refresh(context.TODO(), charmhub.RefreshMany(configs...))
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(refreshResp) != len(applications) {
+		return errors.Errorf("unexpected number of responses %d for applications %d", len(refreshResp), len(applications))
+	}
+	for _, resp := range refreshResp {
+		if err := resp.Error; err != nil && !force {
+			return errors.Annotatef(err, "unable to locate application with series %s", series)
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/facades/client/machinemanager/upgrade_series.go:350]: Makes sense.

File: apiserver/facades/client/machinemanager/upgrade_series.go [hunk 3]
```diff
@@ -217,13 +227,61 @@ func (s upgradeSeriesState) ApplicationsFromMachine(machine Machine) ([]Applicat
 	return results, nil
 }
 
-type upgradeSeriesValidator struct{}
+type upgradeSeriesValidator struct {
+	stateValidator   ApplicationValidator
+	requestValidator ApplicationValidator
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/facades/client/machinemanager/upgrade_series.go:232]: Yeah, that's a good name - IN!

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**